test/test_cubic_b_spline.cpp

```diff
@@ -0,0 +1,217 @@
+    }
+
+    Real step = step_distribution(gen);
+    Real a = dis(gen);
+    boost::math::cubic_b_spline<Real> spline(v.data(), v.size(), a, step);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + a);
+        // This seems like a very large tolerance, but I don't know of any other interpolators
+        // that will be able to do much better on random data.
+        BOOST_CHECK_CLOSE(y, v[i], 10000000*std::numeric_limits<Real>::epsilon());
+    }
+
+}
+
+template<class Real>
+void test_constant_function()
+{
+    std::cout << "Testing that constants are interpolated correctly by cubic b splines on type " << boost::typeindex::type_id<Real>().pretty_name() << "\n";
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    boost::random::uniform_real_distribution<Real> dis(-100, 100);
+    boost::random::uniform_real_distribution<Real> step_distribution(0.0001, 0.1);
+    std::vector<Real> v(500);
+    Real constant = dis(gen);
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        v[i] = constant;
+    }
+
+    Real step = step_distribution(gen);
+    Real a = dis(gen);
+    boost::math::cubic_b_spline<Real> spline(v.data(), v.size(), a, step);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + a);
+        BOOST_CHECK_CLOSE(y, v[i], 10*std::numeric_limits<Real>::epsilon());
+        auto y_prime = spline.interpolate_derivative(i*step + a);
+        BOOST_CHECK_SMALL(y_prime, 5000*std::numeric_limits<Real>::epsilon());
+    }
+
+    // Test that correctly specified left and right-derivatives work properly:
+    spline = boost::math::cubic_b_spline<Real>(v.data(), v.size(), a, step, 0, 0);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + a);
+        BOOST_CHECK_CLOSE(y, v[i], std::numeric_limits<Real>::epsilon());
+        auto y_prime = spline.interpolate_derivative(i*step + a);
+        BOOST_CHECK_SMALL(y_prime, std::numeric_limits<Real>::epsilon());
+    }
+}
+
+
+template<class Real>
+void test_affine_function()
+{
+    std::cout << "Testing that affine functions are interpolated correctly by cubic b splines on type " << boost::typeindex::type_id<Real>().pretty_name() << "\n";
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    boost::random::uniform_real_distribution<Real> dis(-100, 100);
+    boost::random::uniform_real_distribution<Real> step_distribution(0.0001, 0.01);
+    std::vector<Real> v(500);
+    Real a = dis(gen);
+    Real b = dis(gen);
+    Real step = step_distribution(gen);
+    Real x0 = dis(gen);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        v[i] = a*(i*step - x0) + b;
+    }
+
+
+    boost::math::cubic_b_spline<Real> spline(v.data(), v.size(), x0, step);
+
+    for (size_t i = 0; i < v.size(); ++i)
+    {
+        auto y = spline.interpolate_at(i*step + x0);
+        BOOST_CHECK_CLOSE(y, v[i], 10000*std::numeric_limits<Real>::epsilon());
```

Comment:
We can also test in-between the control points since we know what the value is there?